File: framework/languages/java/samples/java/TicTacToe.SessionGateway/src/main/java/systems/zlink/samples/tictactoe/sessiongateway/server/play/handlers/EnsurePlayerActorHandler.java
package systems.zlink.samples.tictactoe.sessiongateway.server.play.handlers;

import java.util.concurrent.CompletionStage;
import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.actors.ZLinkActor;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.samples.tictactoe.sessiongateway.shared.configuration.SampleNames;

public final class EnsurePlayerActorHandler implements ZLinkRequestHandler<String, String> {
    private final ZLinkActorManager actors;

    public EnsurePlayerActorHandler(ZLinkActorManager actors) {
        this.actors = actors;
    }

    @Override
    public CompletionStage<String> handleAsync(String actorId, ZLinkRequestContext context) {
        return actors.getOrCreateAsync(actorId, SampleNames.PlayerActorType)
            .thenCompose(actor -> actor.context()
                .joinEntrySpot(RoutingId.from(SampleNames.EntrySpotRoutingId))
                .submitAsync()
                .thenApply(joined -> snapshot(actor, joined)));
    }

    private static String snapshot(ZLinkActor actor, systems.zlink.framework.actors.ZLinkActorRef ref) {
        return ref.nodeRid().toHex() + "|" + actor.actorId() + "|" + ref.epoch();
    }
}

File: framework/languages/dotnet/samples/Bingo/Server/Play/Adapters/ZLink/Handlers/EnsurePlayerActorHandler.cs
using Systems.Zlink;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Configuration;
using Zlink.Framework.Contracts.Handlers;
using Zlink.Framework.Contracts.Spots;
using Zlink.Framework.Contracts.Streams;
using Zlink.Framework.Contracts.Timers;
using Bingo.Server.Play.Adapters.ZLink.Actors;
using Bingo.Server.Configuration;
using Bingo.Shared.Contracts;

namespace Bingo.Server.Play.Adapters.ZLink.Handlers;

[ZLinkHandlerGroup("play")]
internal sealed class EnsurePlayerActorHandler(
    IZLinkActorManager actors)
    : IZLinkRouteRequestHandler<EnsurePlayerActorReq, EnsurePlayerActorRes>
{
    public async ValueTask<EnsurePlayerActorRes> HandleAsync(
        EnsurePlayerActorReq request,
        ZLinkRouteRequestContext context,
        CancellationToken cancellationToken)
    {
        _ = context;
        var actor = await actors.GetOrCreateAsync(
                request.ActorId,
                SampleNames.PlayerActorType,
                cancellationToken)
            ;
        if (actor is PlayerActor player)
        {
            player.SetDisplayName(request.DisplayName);
        }

        using var entryJoinRequest = Message.From(ReadOnlySpan<byte>.Empty);
        var joined = await actor.Context.JoinEntrySpot(
                RoutingId.From(request.PreferredActorNodeRid),
                entryJoinRequest)
            .Async(cancellationToken)
            ;
        return new EnsurePlayerActorRes
        {
            ActorId = request.ActorId,
            ActorType = SampleNames.PlayerActorType,
            Actor = new ActorRefSnapshot
            {
                NodeRid = joined.Actor.NodeRid.ToString(),
                ActorId = joined.Actor.ActorId,
                Generation = joined.Actor.Generation,
            },
        };
    }
}
